File: packages/dfos-web-relay-go/openapi.go
package relay

/*

  OPENAPI DOCUMENT — the relay's advertised HTTP surface

  WEB-RELAY.md makes serving an OpenAPI document a SHOULD and the well-known's
  `openapi` field the advertisement: a relay that serves one names its URL, and
  absence of the field means none is served. This relay serves one at
  /openapi.json and advertises "/openapi.json" (root-relative, resolved against
  the relay's base URL).

  THE TWO REFERENCE TWINS DIFFER HERE, DELIBERATELY. This relay advertises
  unconditionally; the TS twin advertises only when its caller passes
  createRelay({ openapi: { document } }), and otherwise mounts no route and omits
  the field. Serving is a SHOULD, so both defaults are conformant, and the
  difference is one of packaging rather than protocol: the TS relay is a library
  embedded in someone else's app, where the quiet default is the polite one,
  while this package is what `dfos serve` runs as a deployed relay, where one
  that holds a description of its own surface and declines to say so is just
  less useful. The cost is that relay-conformance/scripts/parity-serve.ts has to opt
  the TS twin in, because the parity gate byte-compares the two well-known
  bodies. Read that opt-in as this note's consequence, not as a workaround for a
  bug.

  THE DOCUMENT IS A COPY, DELIBERATELY. openapi.yaml here is a verbatim copy of
  packages/dfos-web-relay/openapi.yaml, the canonical description of the wire
  surface BOTH reference relays serve. Go's //go:embed cannot reach outside its
  own package directory, so the alternative to a copy is a second, independently
  drifting hand-written document — strictly worse. TestOpenAPIDocumentMatchesCanonical
  is the drift guard: it byte-compares this copy against the canonical file
  (modulo the info.version line, which the release version-sync stamps on the
  canonical only) and names the one-line fix when they diverge.

  THE DOCUMENT DESCRIBES THE FAMILY; THE SERVED COPY DESCRIBES THIS RELAY. The
  canonical file names no `servers`, because there is no host it could name that
  would be true of every relay serving that surface. This relay writes its own
  configured Authority into the served copy's `servers`, so a client reading the
  document resolves operations against THIS relay rather than against whatever
  host the document was authored on. renderOpenAPI carries the full reasoning,
  including why an unconfigured authority correctly yields no `servers` at all.

  YAML in, JSON out: the canonical file is YAML (readable, reviewable, diffable),
  while the served representation is JSON — every OpenAPI consumer reads JSON,
  and the route can then be a plain application/json GET with no content
  negotiation. The conversion runs once, lazily, and is cached.

  The document is DISCOVERY, NEVER AUTHORITY (WEB-RELAY.md): the routes,
  capability gates, and auth rules in the spec govern regardless of what this
  document says. Which is also why the route itself is ungated — a meta surface
  like the well-known, readable before a client knows anything about this relay.

*/

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"gopkg.in/yaml.v3"
)

// openapiPath is the root-relative URL the document is served at and the value
// the well-known advertises in its `openapi` field. One const so the route and
// the advertisement can never disagree.
const openapiPath = "/openapi.json"

//go:embed openapi.yaml
var openapiYAML []byte

// loopbackHosts are the hosts reached in the clear. `api:` surfaces are HTTPS
// surfaces — the CLI refuses to sign a proof for a plaintext request to
// anything else — so the scheme inferred from a bare authority is https for
// every host but these.
var loopbackHosts = map[string]bool{"localhost": true, "127.0.0.1": true, "::1": true, "[::1]": true}

// relayBaseURL is the absolute base URL a request to a relay at authority is
// made against, or "" when there is no authority to describe.
//
// An authority is a bare host or host:port — that is what an identity proof
// binds to — so the scheme has to be inferred. Loopback is served in the clear
// during development; everything else is HTTPS, which is the only scheme a DFOS
// proof may be sent under anyway.
func relayBaseURL(authority string) string {
	if authority == "" {
		return ""
	}
	host := strings.ToLower(authority)
	hostname := host
	if strings.HasPrefix(host, "[") {
		if end := strings.Index(host, "]"); end >= 0 {
			hostname = host[:end+1]
		}
	} else if colon := strings.Index(host, ":"); colon >= 0 {
		hostname = host[:colon]
	}
	if loopbackHosts[hostname] {
		return "http://" + host
	}
	return "https://" + host
}

// renderOpenAPI converts the embedded description to the JSON this relay
// serves. Called once per relay (openapiOnce), never per request.
//
// info.version is overwritten with the running relay's Version rather than the
// canonical file's: the release version-sync stamps the canonical document, not
// this copy, and a served document that reports a version this binary is not is
// worse than no version at all. The well-known's `version` field reports the
// same value, so the two meta surfaces always agree.
//
// SELF-DESCRIPTION. The canonical document carries no `servers` member — it
// describes the surface every DFOS relay serves, not the address of any one of
// them, and a hardcoded host is wrong for every deployment but the one it names.
// The served copy writes `servers` from THIS relay's configured Authority, so a
// client that reads the document reaches this relay. The authority is
// configuration and never the request: Host, X-Forwarded-Host, and the request
// URL are attacker-supplied, and a document that echoed one back would invite a
// client to sign its next proof against a host of the attacker's choosing. With
// no authority configured there is nothing honest to write, so the member stays
// absent — and OpenAPI then resolves operations against the URL the document was
// retrieved from, which for a self-served document is this relay either way.
func renderOpenAPI(authority string) ([]byte, error) {
	var parsed any
	if err := yaml.Unmarshal(openapiYAML, &parsed); err != nil {
		return nil, err
	}
	document := jsonifyYAML(parsed)
	root, ok := document.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("openapi document is not a mapping")
	}
	if info, ok := root["info"].(map[string]any); ok {
		info["version"] = Version
	}
	if base := relayBaseURL(authority); base != "" {
		root["servers"] = []any{map[string]any{"url": base}}
	} else {
		delete(root, "servers")
	}
	return json.Marshal(document)
}

// openapiDocument returns the description this relay serves, converting once.
func (r *Relay) openapiDocument() ([]byte, error) {
	r.openapiOnce.Do(func() {
		r.openapiJSON, r.openapiErr = renderOpenAPI(r.authority)
	})
	return r.openapiJSON, r.openapiErr
}

// jsonifyYAML makes a decoded YAML value JSON-marshalable. yaml.v3 decodes a
// mapping into map[string]any when every key is a string and falls back to
// map[any]any otherwise — and encoding/json cannot marshal the latter. Rewriting
// the fallback keys with fmt.Sprint keeps a non-string key (an unquoted `200:`
// response code, say) from turning the whole document into a 500.
func jsonifyYAML(value any) any {
	switch typed := value.(type) {
	case map[string]any:
		for key, child := range typed {
			typed[key] = jsonifyYAML(child)
		}
		return typed
	case map[any]any:
		converted := make(map[string]any, len(typed))
		for key, child := range typed {
			converted[fmt.Sprint(key)] = jsonifyYAML(child)
		}
		return converted
	case []any:
		for i, child := range typed {
			typed[i] = jsonifyYAML(child)
		}
		return typed
	default:
		return value
	}
}

// handleOpenAPI serves the relay's OpenAPI document. Ungated: a meta surface,
// like the well-known that advertises it.
func (r *Relay) handleOpenAPI(w http.ResponseWriter, _ *http.Request) {
	document, err := r.openapiDocument()
	if err != nil {
		writeError(w, 500, "openapi document unavailable")
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(200)
	_, _ = w.Write(document)
}

File: packages/dfos-web-relay/src/openapi.ts
/*

  SELF-DESCRIPTION — the `servers` a served OpenAPI document names

  The canonical openapi.yaml carries NO `servers` member. It describes the wire
  surface every DFOS relay serves, not the address of any one of them, and a
  document that hardcodes one host is wrong everywhere else: a client that reads
  `http://localhost:3000` out of a document fetched from relay.example.com will
  resolve every operation against localhost and reach nothing. (That is not
  hypothetical — it is the bug this module exists to close.)

  OpenAPI's own default is the correct behavior for a self-served document: with
  `servers` absent, operations resolve against the URL the document was
  retrieved from. A relay that serves its description at /openapi.json is
  therefore already right by saying nothing. This module exists for the strictly
  better case — a relay that KNOWS its own authority can say so, and then the
  document stays right after it is saved to disk, mirrored, or handed to a tool
  that lost track of where it came from.

  THE AUTHORITY IS CONFIGURATION, NEVER THE REQUEST. `servers` is written from
  `RelayOptions.authority` — the same host binding every identity proof is
  checked against — and never from Host, X-Forwarded-Host, or the request URL.
  Those are attacker-supplied, and a document that echoed one back would invite
  a client to sign its next proof against a host of the attacker's choosing.
  When no authority is configured there is nothing honest to write, so the
  member is left absent and OpenAPI's default does the work.

  THE DOCUMENT ALSO DESCRIBES ITSELF, AND THAT ENTRY MOVES TOO. The canonical
  document carries a `/openapi.json` path entry for the operation that returns
  the document — so a deployment that serves it at some other route (a relay
  mounted behind a gateway, a house convention of `/schema.json`) advertised one
  path in its well-known while the served document claimed the document lived at
  another, which 404s. The `servers` rewrite alone did not close that: it fixes
  the ORIGIN every operation resolves against, not the PATH of this one. So the
  self-entry is rekeyed to the route the document is actually served at,
  in place, with its contents untouched and no other path moved.

*/

/**
 * The route the canonical document describes ITSELF at — the `paths` key of its
 * `getOpenApiDocument` operation, and the route `createRelay` serves at when a
 * deployment configures none. One const, so the served copy and the route
 * validation can never disagree about which entry is the self-entry.
 */
export const DEFAULT_OPENAPI_ROUTE = '/openapi.json';

/**
 * Hosts that are reached in the clear. `api:` surfaces are HTTPS surfaces — the
 * CLI refuses to sign a proof for a plaintext request to anything else — so the
 * scheme this module infers from a bare authority is `https` for every host but
 * these.
 */
const LOOPBACK_HOSTS = new Set(['localhost', '127.0.0.1', '::1', '[::1]']);

/**
 * The absolute base URL a request to a relay at `authority` is made against.
 *
 * An authority is a bare `host` or `host:port` (that is what a proof binds to),
 * so the scheme has to be inferred. Loopback is served in the clear during
 * development; everything else is HTTPS, which is the only scheme a DFOS
 * request proof may be sent under anyway.
 *
 * Returns `undefined` when there is no authority to describe.
 */
export const relayBaseUrl = (authority: string | undefined): string | undefined => {
  if (authority === undefined || authority === '') return undefined;
  const host = authority.toLowerCase();
  const hostname = host.startsWith('[')
    ? host.slice(0, host.indexOf(']') + 1)
    : (host.split(':')[0] ?? '');
  return `${LOOPBACK_HOSTS.has(hostname) ? 'http' : 'https'}://${host}`;
};

/**
 * Rekey the document's own path entry to the route it is actually served at.
 *
 * PURE, and deliberately conservative — it fires only when the document says
 * something FALSE about itself. Two cases leave the paths untouched because in
 * both the document is already saying something intentional about `servedRoute`:
 * a document with no default self-entry never described itself here, and one
 * that ALREADY carries an entry at `servedRoute` would have it clobbered.
 *
 * The entry is rekeyed IN PLACE: it keeps its position in the path list and its
 * value by reference, so nothing but the key changes and no other path moves.
 */
const relocateSelfPath = (
  served: Record<string, unknown>,
  servedRoute: string,
): Record<string, unknown> => {
  if (servedRoute === DEFAULT_OPENAPI_ROUTE) return served;
  const paths = served['paths'];
  if (typeof paths !== 'object' || paths === null || Array.isArray(paths)) return served;
  const entries = Object.entries(paths as Record<string, unknown>);
  if (!entries.some(([path]) => path === DEFAULT_OPENAPI_ROUTE)) return served;
  if (entries.some(([path]) => path === servedRoute)) return served;
  return {
    ...served,
    paths: Object.fromEntries(
      entries.map(([path, item]) => [path === DEFAULT_OPENAPI_ROUTE ? servedRoute : path, item]),
    ),
  };
};

/**
 * The document as this relay serves it: a copy of `document` whose `servers`
 * names this relay's own authority — or, when no authority is configured, a copy
 * with no `servers` member at all — and whose self-describing path entry names
 * `servedRoute` rather than the default.
 *
 * A copy, not a mutation: the caller's document is very often the shared import
 * of `@metalabel/dfos-web-relay/openapi.json`, and two relays in one process
 * would otherwise overwrite each other's self-description. Only the top level is
 * cloned — plus `paths`, and only when the self-entry actually moves.
 */
export const selfDescribingDocument = (
  document: unknown,
  authority: string | undefined,
  servedRoute: string = DEFAULT_OPENAPI_ROUTE,
): Record<string, unknown> => {
  const source = document as Record<string, unknown>;
  const served = { ...source };
  const baseUrl = relayBaseUrl(authority);
  if (baseUrl === undefined) {
    delete served['servers'];
  } else {
    served['servers'] = [{ url: baseUrl }];
  }
  return relocateSelfPath(served, servedRoute);
};
